File: src/main/java/com/example/tutorial/common/datamodel/Task.java
package com.example.tutorial.common.datamodel;

import lombok.*;

/**
 * Task data model for producer-consumer and concurrent processing patterns.
 * Used in: ProducerConsumerPattern, ThreadPoolExecutorDemo
 */
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Task {

  public enum Priority {
    LOW(1),
    MEDIUM(2),
    HIGH(3),
    CRITICAL(4);

    private final int level;

    Priority(int level) {
      this.level = level;
    }

    public int getLevel() {
      return level;
    }
  }

  public enum Status {
    PENDING, IN_PROGRESS, COMPLETED, FAILED
  }

  private String taskId;
  private String name;
  private String description;
  private Priority priority;
  private Status status;
  private long createdAt;
  private long completedAt;

  /**
   * Creates a new pending task.
   */
  public static Task of(String name, Priority priority) {
    return Task.builder()
        .taskId("TASK-" + System.nanoTime())
        .name(name)
        .priority(priority)
        .status(Status.PENDING)
        .createdAt(System.currentTimeMillis())
        .build();
  }

  /**
   * Creates a simple task with medium priority.
   */
  public static Task of(String name) {
    return of(name, Priority.MEDIUM);
  }

  /**
   * Marks task as completed.
   */
  public void complete() {
    this.status = Status.COMPLETED;
    this.completedAt = System.currentTimeMillis();
  }

  @Override
  public String toString() {
    return String.format("[%s] %s (%s)", priority, name, status);
  }
}

File: src/main/python/com/example/tutorial/common/datamodel/task.py
"""
Task data model for producer-consumer and concurrent processing patterns.
Used in: ProducerConsumerPattern, ThreadPoolExecutorDemo
"""

from dataclasses import dataclass, field
from enum import Enum
from typing import Optional
import time


class Priority(Enum):
    """Task priority enumeration."""
    LOW = 1
    MEDIUM = 2
    HIGH = 3
    CRITICAL = 4


class Status(Enum):
    """Task status enumeration."""
    PENDING = "PENDING"
    IN_PROGRESS = "IN_PROGRESS"
    COMPLETED = "COMPLETED"
    FAILED = "FAILED"


@dataclass
class Task:
    """Task data model for producer-consumer patterns."""
    task_id: str
    name: str
    priority: Priority
    status: Status = Status.PENDING
    description: Optional[str] = None
    created_at: float = field(default_factory=time.time)
    completed_at: Optional[float] = None

    @staticmethod
    def of(name: str, priority: Priority = Priority.MEDIUM) -> 'Task':
        """Creates a new pending task."""
        return Task(
            task_id=f"TASK-{int(time.time() * 1000000)}",
            name=name,
            priority=priority,
            status=Status.PENDING,
            created_at=time.time()
        )

    def complete(self) -> None:
        """Marks task as completed."""
        self.status = Status.COMPLETED
        self.completed_at = time.time()

    def __str__(self) -> str:
        return f"[{self.priority.name}] {self.name} ({self.status.name})"
